ESQL: Fix ReplaceMissingFieldsWithNull

alex-spies opened the pull request:
Fix https://github.com/elastic/elasticsearch/issues/126030
Fix https://github.com/elastic/elasticsearch/issues/126036
Fix https://github.com/elastic/elasticsearch/issues/121754

There are multiple problems with `ReplaceMissingFieldsWithNull`:

When encountering projections, it tries to do the job of field extraction for missing fields by injecting an Eval that creates a literal null with the same name id as the field attribute for the missing field. But:
1. We only insert an Eval in case that a Project relies on the missing attribute. There could be other plan nodes that rely on the missing attribute.
2. Even for Projects, we only insert an Eval in case we squarely project for the field - in case of aliases (e.g. from RENAME), we do nothing.
3. In case of multiple Projects that use this attribute, we create multiple attributes with the original field attribute's id, causing a wrong Layout. This triggered https://github.com/elastic/elasticsearch/issues/121754. **This could even happen if there's simply a command upstream from a single Project in case the upstream command triggers a field extraction for the missing field**.

There are multiple ways of dealing with this:
1. simple solutions:
   1. We could move the creation of the Evals into `InsertFieldExtraction`. This will be correct and simple, but we will miss out on important logical optimizations, namely turning the `EsRelation` into a `LocalRelation` in case all fields are missing, or propagating the literal nulls created in place of the missing field (enabling more constant folding etc.).
   2. We could just place an Eval directly on top of any EsRelation that has missing fields, introducing literal nulls in place of the missing attributes early on. This has a risk of emitting null blocks unnecessarily early in the execution, but the logical optimizer should be able to deal with this.
2. more complex solutions:
   1. We could just avoid replacing the same field attribute twice. This does not address the problem that upstream query plan nodes might trigger a field extraction, leading to duplicate name ids again.
   2. We could keep the original logic but just use new attribute ids. This is what my first attempt (https://github.com/elastic/elasticsearch/pull/125656) does; it **seems** to work fine, but downstream nodes need updating to use the Eval's newly created ReferenceAttributes instead of the original FieldAttributes, at least in terms of the name id of the ReferenceAttribute. **But! We cannot just replace field attributes by reference attributes** in downstream nodes, that can lead to class cast exceptions...
   3. Instead of injecting Evals just before Projects, we could inject them just before the first query plan node that needs them. This would replicate the logic in `InsertFieldExtraction`, but in the local _logical_ optimizer. In addition to the duplicated logic, this has a risk of accidentally messing up the ordering of the output fields (Evals place their attributes always on the right hand side), which could cause serious problems unless there's a Project/Stats somewhere downstream (which is probably always given, but still).

This PR uses approach 1.ii. (Approach 1.i would also be possible and is arguably more natural. It would require double checking that we don't miss out on some optimizations, though.)

alex-spies committed:
Revert changes to Layout.java

The change in 80125a4bac31f54cc86fdee26ef0ac5dfa61eee3 is a quick fix
and allows breaking an invariant of Layout. Revert that.

alex-spies committed:
Simplify ReplaceMissingFieldWithNull

When encountering projections, it tries to do the job of field
extraction for missing fields by injecting an Eval that creates a
literal null with the same name id as the field attribute for the
missing field. This is wrong:
1. We only insert an Eval in case that a Project relies on the missing
   attribute. There could be other plan nodes that rely on the missing
   attribute.
2. Even for Projects, we only insert an Eval in case we squarely project
   for the field - in case of aliases (e.g. from RENAME), we do nothing.
3. In case of multiple Projects that use this attribute, we create
   multiple attributes with the original field attribute's id, causing
   a wrong Layout. This triggered
   https://github.com/elastic/elasticsearch/issues/121754.

alex-spies committed:
Revive logic for EsRelation instead of Project

alex-spies committed:
Update LocalLogicalPlanOptimizerTests

alex-spies committed:
Update docs/changelog/125764.yaml

elasticsearchmachine commented:
Hi @alex-spies, I've created a changelog YAML for you.

alex-spies committed:
Update test expectations

fang-xing-esql reviewed (commented):
Thank you @alex-spies, I added a couple of things that I can think of.

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/local/ReplaceMissingFieldWithNull.java:
```diff
@@ -43,84 +41,79 @@ public class ReplaceMissingFieldWithNull extends ParameterizedRule<LogicalPlan,
-                    // TODO: Should do a searchStats lookup for join attributes instead of just ignoring them here
-                    // See TransportSearchShardsAction
+    private LogicalPlan missingToNull(LogicalPlan plan, Predicate<FieldAttribute> shouldBeRetained) {
+        if (plan instanceof EsRelation relation) {
```
I tried your suggestion before arriving at the current solution. It'd be nicer, in fact, than going by the index mode and looking for LOOKUP.

However, this clashes with how we represent `LookupJoinExec`, which unfortunately uses a fragment on the right which contains an `EsRelation`.

Example: consider this csv test.
```
from * | lookup join languages_lookup on language_code
```
During the physical planning stage, we will at one point have
```
OutputExec[org.elasticsearch.xpack.esql.plugin.ComputeService$$Lambda/0x000000005e684a38@65a57e36]
\_LimitExec[1000[INTEGER]]
  \_LookupJoinExec[[language_code{f}#22],[language_code{f}#113],[language_name{f}#114]]
    |_LimitExec[1000[INTEGER]]
    | \_ExchangeSourceExec[[...],false]
    \_FragmentExec[filter=null, estimatedRowSize=0, reducer=[], fragment=[<>
EsRelation[languages_lookup][LOOKUP][language_code{f}#113, language_name{f}#114]<>]]
```
Unfortunately, in [PlannerUtils.localPlan](https://github.com/elastic/elasticsearch/blob/1d6c6a59c86de232c65a29579a12e0fdb0eb97ad/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/PlannerUtils.java#L163), we descend into all fragments, separately, and optimize them, triggering `ReplaceMissingFieldsWithNulls`.

This means that we have to look for the `LOOKUP` index mode right now if we don't want to increase the scope of this PR. And even if we increased the scope, I'm not sure what the right solution would be. (But I have a hunch that the right hand side of `LookupJoinExec` shouldn't be a fragment.)

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/local/ReplaceMissingFieldWithNull.java:
```diff
@@ -43,84 +41,79 @@ public class ReplaceMissingFieldWithNull extends ParameterizedRule<LogicalPlan,
+        // Do not use the attribute name, this can deviate from the field name for union types; use fieldName() instead.
+        // Also retain fields from lookup indices because we do not have stats for these.
+        Predicate<FieldAttribute> shouldBeRetained = f -> f.field() instanceof PotentiallyUnmappedKeywordEsField
+            || (localLogicalOptimizerContext.searchStats().exists(f.fieldName()) || lookupFields.contains(f));
```
I thought the same initially, but `lookupFields` is actually not a proper subset of `joinAttributes`. The reason for this is that we only mark indices used for lookup with mode `LOOKUP`. ("Index mode" is probably a misnomer at this point, it should be something like "intended use".)

`EsRelation`s that originate from the `FROM` command cannot have `LOOKUP` mode. In fact, the [only line of code](https://github.com/elastic/elasticsearch/blob/1d6c6a59c86de232c65a29579a12e0fdb0eb97ad/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/parser/LogicalPlanBuilder.java#L585) where esql ever assigns this mode is when we parse a `LOOKUP JOIN`.

This can also be checked by debugging the following queries against the csv test data (which I did for good measure):
```
from languages_lookup,employees | lookup join languages_lookup on language_code

from languages_lookup | lookup join languages_lookup on language_code
```

alex-spies commented:
Thanks for the remarks @fang-xing-esql . I'll add a couple of comments to the code so that we don't have to scratch our heads with the same questions in the future.

alex-spies committed:
Do not prune attributes from EsRelation

This can lead to empty output, which leads to the EsRelation being
replaced by a LocalRelation with 0 rows.

alex-spies committed:
Merge remote-tracking branch 'upstream/main' into fix-replace-missing-fields-with-null

alex-spies committed:
Add tests + capability

elasticsearchmachine commented:
Hi @alex-spies, I've updated the changelog YAML for you.

alex-spies committed:
Update docs/changelog/125764.yaml

alex-spies committed:
Add comments

elasticsearchmachine commented:
Pinging @elastic/es-analytical-engine (Team:Analytics)

costin reviewed (approved):
LGTM

fang-xing-esql reviewed (approved):
LGTM, thank you @alex-spies !

alex-spies committed:
Merge remote-tracking branch 'upstream/main' into fix-replace-missing-fields-with-null

alex-spies committed:
Merge remote-tracking branch 'upstream/main' into fix-replace-missing-fields-with-null

alex-spies commented:
Last CI run was green except for elasticsearch-ci/part-1 timing out:
```
2025-04-02 20:08:37 CEST	> Task :build-tools:reaper:spotlessJava
2025-04-03 01:07:06 CEST	# Received cancellation signal, interrupting
2025-04-03 01:07:06 CEST	🚨 Error: The command was interrupted by a signal: signal: terminated
```

Restarted CI as there was a merge conflict now.

alex-spies commented:
![image](https://github.com/user-attachments/assets/b99f77fb-780d-4b82-a206-b98697b6fc22)
The checks here complain that they're still waiting on elasticsearch-ci/part-2, but in buildkite, this pipeline already passed green.

alex-spies commented:
![image](https://github.com/user-attachments/assets/7e0afc5c-bb18-4de5-8adb-fa87abde2bcf)
Seems like the reporting of the status of `elasticsearch-ci/part-2` may be stuck.

But since the CI passed in CI, this is safe to merge.

alex-spies commented:
Ah yeah, publishing the build scan failed:
```
BUILD SUCCESSFUL in 36m 35s
3025 actionable tasks: 1247 executed, 1652 from cache, 126 up-to-date
Publishing build scan...
Publishing build scan failed due to network error 'java.net.SocketTimeoutException: Read timed out' (2 retries remaining)...
Publishing build scan failed due to network error 'java.net.SocketTimeoutException: Read timed out' (1 retry remaining)...
A network error occurred.
If you require assistance with this problem, please report it to your Develocity administrator and include the following information via copy/paste.
----------
Gradle version: 8.13
Plugin version: 3.19.2
Request URL: https://gradle-enterprise.elastic.co/scans/publish/gradle/3.19.2/token
Request ID: dea4cbad-e35c-4b39-af24-a35bcbaf4292
Exception: java.net.SocketTimeoutException: Read timed out
----------
```
